src/ucs/datastruct/linear_func.h

```diff
@@ -57,14 +57,100 @@ ucs_linear_func_apply(ucs_linear_func_t f, double x)
 /**
  * Sum two linear functions.
  *
- * @param [out] result   Filled with the resulting linear function.
  * @param [in]  func1    First function to add.
  * @param [in]  func2    Second function to add.
+ *
+ * @return Linear function representing (func1 + func2)
  */
 static UCS_F_ALWAYS_INLINE ucs_linear_func_t
 ucs_linear_func_add(ucs_linear_func_t func1, ucs_linear_func_t func2)
 {
     return ucs_linear_func_make(func1.c + func2.c, func1.m + func2.m);
 }
 
+
+/**
+ * Subtract one linear function from another.
+ *
+ * @param [in]  func1    Linear function to subtract from.
+ * @param [in]  func2    Linear function to subtract.
+ *
+ * @return Linear function representing (func1 - func2)
+ */
+static inline ucs_linear_func_t
+ucs_linear_func_sub(ucs_linear_func_t func1, ucs_linear_func_t func2)
+{
+    return ucs_linear_func_make(func1.c - func2.c, func1.m - func2.m);
+}
+
+
+/**
+ * Add one linear function to another in-place.
+ *
+ * @param [inout]  func1    First linear function to add, and the result of the
+ *                          operation
+ * @param [in]     func2    Second function to add.
+ */
+static inline void
+ucs_linear_func_add_inplace(ucs_linear_func_t *func1, ucs_linear_func_t func2)
+{
+    func1->m += func2.m;
+    func1->c += func2.c;
+}
+
+
+/**
+ * Substitute the "x" argument of a linear function by another linear function
+ * and return the composition of the functions.
+ *
+ * @param [in]  outer   Linear function whose "x" argument should be substituted
+ * @param [in]  inner   Linear function to substitute.
+ *
+ * @return Linear function representing outer(inner(x))
+ */
+static inline ucs_linear_func_t
+ucs_linear_func_compose(ucs_linear_func_t outer, ucs_linear_func_t inner)
+{
+    /*
+     * let  outer(x) = outer.m*x + outer.c, and g(x) = g.m*x + g.c
+     * then outer(g(x)) = outer.m(inner.m*x + g.c) + outer.c =
+     *                    (outer.m * inner.m)x + (outer.m*g.c + outer.c)
+     */
+    return ucs_linear_func_make((outer.m * inner.c) + outer.c,
+                                outer.m * inner.m);
+}
+
+
+/**
+ * Find the intersection point between two linear functions. If the functions
+ * do not intersect, the result is undefined.
```

Comment:
maybe return the UCS status (to report that `m1 - m2 == 0`) instead of the X-value?
otherwise, it would be an arithmetic error (division by 0)

Reply:
added check for nan/inf, actually float division by zero does not cause error, but returns NaN.

Reply:
> actually float division by zero does not cause error, but returns NaN.

yes
btw, @brminich @yosefe maybe boolean value would be better than UCS status, wdyt? 0 - not intersected, 1 - intersected